File: leetcode/543_Diameter_of_Binary_Tree.cpp
#include <bits/stdc++.h>
using namespace std;
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 };

class Solution {
public:
int dia=0;
    int diameterOfBinaryTree(TreeNode* root) {
        func(root);
        return dia;
    }
    int func(TreeNode* root){
        if(root==0)return 0;
        int left_height=func(root->left);
        int right_height=func(root->right);
        dia=max(dia, left_height+right_height);
        return max(left_height, right_height)+1;
    }
};
int main(){
    Solution s;
    TreeNode* root=new TreeNode(1);
    root->left=new TreeNode(2);
    root->right=new TreeNode(3);
    root->left->left=new TreeNode(4);
    root->left->right=new TreeNode(5);
    cout<<s.diameterOfBinaryTree(root)<<endl;
    return 0;
}

File: leetcode/543_Diameter_of_Binary_Tree.java
import java.util.*;
public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}
class Solution {
    int dia;
    public int func(TreeNode root){
        if(root==null)return 0;
        int lefth=func(root.left);
        int righth=func(root.right);
        dia=Math.max(dia,lefth+righth);
        return Math.max(lefth,righth)+1;
    }
    public int diameterOfBinaryTree(TreeNode root) {
        func(root);
        return dia;
    }
}
public class Main {
    public static void main(String[] args) {
        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.right = new TreeNode(3);
        root.left.left = new TreeNode(4);
        root.left.right = new TreeNode(5);

        Solution solution = new Solution();
        int diameter = solution.diameterOfBinaryTree(root);
        System.out.println("Diameter of the binary tree: " + diameter);
    }
}